### DISCUSSION TITLE
Shader binding and matrix calculations inside loop.

### DISCUSSION THREAD

**Discussion by ghost** [raysan5/raylib#2062] [Q&A] [answered]
Hello again.

This concerns the block of code at: https://github.com/raysan5/raylib/blob/226c0e362c6784edf7e86e94624da655d185b259/src/rlgl.h#L2380

```
           // Set current shader and upload current MVP matrix
            glUseProgram(RLGL.State.currentShaderId);

            // Create modelview-projection matrix and upload to shader
            Matrix matMVP = rlMatrixMultiply(RLGL.State.modelview, RLGL.State.projection);
            float matMVPfloat[16] = {
                matMVP.m0, matMVP.m1, matMVP.m2, matMVP.m3,
                matMVP.m4, matMVP.m5, matMVP.m6, matMVP.m7,
                matMVP.m8, matMVP.m9, matMVP.m10, matMVP.m11,
                matMVP.m12, matMVP.m13, matMVP.m14, matMVP.m15
            };
            glUniformMatrix4fv(RLGL.State.currentShaderLocs[RL_SHADER_LOC_MATRIX_MVP], 1, false, matMVPfloat);
```

This code is run inside a loop, eventhough none of the values involved rely on anything defined in the loop.
This means this code can be moved out of the loop and it will still work as intended.

Also not a bug, and outside of Stereo rendering, this won't have any effect. But when doing stereo rendering, we are binding the shader program and computing/uploading the MVP matrix twice without any benefit.

**Comment by ghost** [comment]
Some more uniforms are set a few lines down at:

https://github.com/raysan5/raylib/blob/226c0e362c6784edf7e86e94624da655d185b259/src/rlgl.h#L2414

These can also be moved out of the loop.

**Comment by raysan5** [comment, marked as answer]
@hexaflex Actually `RLGL.State.modelview` and `RLGL.State.projection` depend on:
```c
    // Set current eye view offset to modelview matrix
    rlSetMatrixModelview(rlMatrixMultiply(matModelView, RLGL.State.viewOffsetStereo[eye]));
    // Set current eye projection matrix
    rlSetMatrixProjection(RLGL.State.projectionStereo[eye]);
```
In any case, the overhead of calling those functions for one or two eyes is minimal so I prefer to keep all the code inside the loop.

**Reply by ghost** [reply]
> `RLGL.State.modelview` and `RLGL.State.projection` depend on ...

I cant believe I missed that. Thanks.
